pkg/controller/v1beta1/inferenceservice/reconcilers/ingress/rawingress_reconciler.go

```diff
@@ -0,0 +1,231 @@
+								Name: componentName,
+								Port: netv1.ServiceBackendPort{
+									Number: constants.CommonDefaultHttpPort,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	return rule
+}
+
+func generateMetadata(isvc *v1beta1api.InferenceService,
+	componentType constants.InferenceServiceComponent) metav1.ObjectMeta {
+	var name string
+	switch componentType {
+	case constants.Transformer:
+		name = constants.DefaultTransformerServiceName(isvc.Name)
+	case constants.Explainer:
+		name = constants.DefaultExplainerServiceName(isvc.Name)
+	case constants.Predictor:
+		name = constants.DefaultPredictorServiceName(isvc.Name)
+	}
+	//get annotations from isvc
+	annotations := utils.Filter(isvc.Annotations, func(key string) bool {
+		return !utils.Includes(constants.ServiceAnnotationDisallowedList, key)
+	})
+	objectMeta := metav1.ObjectMeta{
+		Name:      name,
+		Namespace: isvc.Namespace,
+		Labels: utils.Union(isvc.Labels, map[string]string{
+			constants.InferenceServicePodLabelKey: isvc.Name,
+			constants.KServiceComponentLabel:      string(componentType),
+		}),
+		Annotations: annotations,
+	}
+	return objectMeta
+}
+
+//GenerateIngressHost return the config domain in configmap.IngressDomain
+func GenerateIngressHost(ingressConfig *v1beta1api.IngressConfig,
+	isvc *v1beta1api.InferenceService,
+	componentType string,
+	topLevelFlag bool) string {
+	if ingressConfig.IngressDomain == "" {
+		ingressConfig.IngressDomain = "example.com"
+	}
+	metadata := generateMetadata(isvc, constants.InferenceServiceComponent(componentType))
+	if topLevelFlag {
+		return metadata.Name + "-" + metadata.Namespace + "." + ingressConfig.IngressDomain
+	} else {
+		return isvc.Name + "-" + isvc.Namespace + "." + ingressConfig.IngressDomain
+	}
+}
+
+func createRawIngress(client client.Client,
+	scheme *runtime.Scheme,
+	isvc *v1beta1api.InferenceService,
+	ingressConfig *v1beta1api.IngressConfig) (*netv1.Ingress, error) {
+	topLevelFlag := false
+	var rules []netv1.IngressRule
+	if isvc.Spec.Transformer != nil {
+		host := GenerateIngressHost(ingressConfig, isvc, string(constants.Transformer), topLevelFlag)
+		//set topLevelFlag as true
+		topLevelFlag = true
+		podSpec := corev1.PodSpec(isvc.Spec.Transformer.PodSpec)
+		rules = append(rules, generateRule(&podSpec, host, constants.DefaultTransformerServiceName(isvc.Name)))
+	}
+	if isvc.Spec.Explainer != nil {
+		host := GenerateIngressHost(ingressConfig, isvc, string(constants.Explainer), topLevelFlag)
+		//set topLevelFlag as true
+		topLevelFlag = true
+		podSpec := corev1.PodSpec(isvc.Spec.Explainer.PodSpec)
+		rules = append(rules, generateRule(&podSpec, host, constants.DefaultExplainerServiceName(isvc.Name)))
+	}
+	//add predictor rule
+	host := GenerateIngressHost(ingressConfig, isvc, string(constants.Predictor), topLevelFlag)
+	podSpec := corev1.PodSpec(isvc.Spec.Predictor.PodSpec)
+	rules = append(rules, generateRule(&podSpec, host, constants.DefaultPredictorServiceName(isvc.Name)))
```

Comment:
How does this work when there are transformer and predictor, I think it creates the conflicting rules right?

Reply:
the host is different, so there is no conflicting.